### PR TITLE
Online marglik as a method

### PR DESCRIPTION
**Pull request by AlexImmer**

### CODE REVIEW

File: laplace/marglik.py [hunk 1]
```diff
@@ -0,0 +1,248 @@
...
+            else:
+                crit_factor = temperature
+            prior_prec = torch.exp(log_prior_prec).detach()
+            theta = parameters_to_vector(model.parameters())
+            delta = expand_prior_precision(prior_prec, model)
+            f = model(X)
+            loss = criterion(f, y) + (0.5 * (delta * theta) @ theta) / N / crit_factor
+            loss.backward()
+            optimizer.step()
+            epoch_loss += loss.cpu().item() / len(train_loader)
```
Comment (runame) [inline laplace/marglik.py:189]: The loss is not exact since the size of the last batch is typically smaller than the batch size, i.e. the correct thing would be to multiply the loss of every batch by `len(X)` and then devide by `N` at the end.

File: laplace/marglik.py [hunk 2]
```diff
@@ -0,0 +1,248 @@
...
+def marglik_training(
+    model,
+    train_loader,
+    likelihood='classification',
+    optimizer_cls=Adam,
+    optimizer_kwargs=None,
+    scheduler_cls=None,
+    scheduler_kwargs=None,
+    n_epochs=300,
+    laplace=KronLaplace,
```
Comment (runame) [inline laplace/marglik.py:23]: Maybe it makes sense to use the `Laplace` function interface, since we use it everywhere else as the UI. So here you would pass the `laplace_kwargs` or `subset_of_weights`, `hessian_structure`, and `laplace_kwargs`.

**Review by runame — COMMENTED** [review]
Looks good! I think my first comment below regarding the `Laplace` interface is the only thing to discuss.

One more very minor question: any reason why you did not call the file `marglik_training.py`?

**Comment by runame** [comment]
Not sure what we decided before, but we also considered having a designated directory for methods, right? We can still add that later when we have more methods though.

File: laplace/marglik.py (hunk 1 above)
Comment (AlexImmer) [reply laplace/marglik.py]: Good catch, thanks. I fixed it.

File: laplace/marglik_training.py [hunk 3]
```diff
@@ -0,0 +1,259 @@
...
+            else:
+                crit_factor = temperature
+            prior_prec = torch.exp(log_prior_prec).detach()
+            theta = parameters_to_vector(model.parameters())
+            delta = expand_prior_precision(prior_prec, model)
+            f = model(X)
+            loss = criterion(f, y) + (0.5 * (delta * theta) @ theta) / N / crit_factor
+            loss.backward()
+            optimizer.step()
+            epoch_loss += loss.cpu().item() / len(train_loader)
```
Comment (runame) [inline laplace/marglik_training.py:194]: If I'm not missing something you now devide the loss by `N` in line 202, but you still have to multiply by `len(X)` here instead of divide by `len(train_loader)`.

File: laplace/marglik.py (hunk 2 above)
Comment (AlexImmer) [reply laplace/marglik.py]: Done as suggested, thanks.

File: laplace/marglik_training.py (hunk 3 above)
Comment (AlexImmer) [reply laplace/marglik_training.py:194]: Oh yes, thanks. I mixed it up with `epoch_perf`...

**Review by runame — APPROVED** [review]
